File: problem/ai-battle-series/07-no-forbidden/src/main.py
import sys


def calc_digit_sum(v):
    n = -v if v < 0 else v
    if n == 0:
        return 0
    total = 0
    while n > 0:
        total += n % 10
        n //= 10
    return total


def less_item(a, b):
    if a[1] != b[1]:
        return a[1] < b[1]
    return a[0] < b[0]


def merge_step(arr, temp, left, mid, right):
    i = left
    j = mid
    k = left
    while i < mid and j < right:
        if less_item(arr[i], arr[j]):
            temp[k] = arr[i]
            i += 1
        else:
            temp[k] = arr[j]
            j += 1
        k += 1
    while i < mid:
        temp[k] = arr[i]
        i += 1
        k += 1
    while j < right:
        temp[k] = arr[j]
        j += 1
        k += 1
    for idx in range(left, right):
        arr[idx] = temp[idx]


def merge_order(arr, temp, left, right):
    if right - left <= 1:
        return
    mid = left + (right - left) // 2
    merge_order(arr, temp, left, mid)
    merge_order(arr, temp, mid, right)
    merge_step(arr, temp, left, mid, right)


def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    n = int(next(it))
    values = [int(next(it)) for _ in range(n)]
    arr = [(v, calc_digit_sum(v)) for v in values]
    temp = arr.copy()
    merge_order(arr, temp, 0, len(arr))
    sys.stdout.write(" ".join(str(v) for v, _ in arr) + "\n")


if __name__ == "__main__":
    main()

File: problem/ai-battle-series/07-no-forbidden/src/main.cpp
#include <iostream>
#include <vector>

using namespace std;

struct Item {
    int value;
    int digit_sum;
};

static int calc_digit_sum(int v) {
    int n = v < 0 ? -v : v;
    int sum = 0;
    if (n == 0) {
        return 0;
    }
    while (n > 0) {
        sum += n % 10;
        n /= 10;
    }
    return sum;
}

static bool less_item(const Item &a, const Item &b) {
    if (a.digit_sum != b.digit_sum) {
        return a.digit_sum < b.digit_sum;
    }
    return a.value < b.value;
}

static void merge_step(vector<Item> &arr, vector<Item> &tmp, int left, int mid, int right) {
    int i = left;
    int j = mid;
    int k = left;
    while (i < mid && j < right) {
        if (less_item(arr[i], arr[j])) {
            tmp[k++] = arr[i++];
        } else {
            tmp[k++] = arr[j++];
        }
    }
    while (i < mid) {
        tmp[k++] = arr[i++];
    }
    while (j < right) {
        tmp[k++] = arr[j++];
    }
    for (i = left; i < right; i++) {
        arr[i] = tmp[i];
    }
}

static void merge_order(vector<Item> &arr, vector<Item> &tmp, int left, int right) {
    if (right - left <= 1) {
        return;
    }
    int mid = left + (right - left) / 2;
    merge_order(arr, tmp, left, mid);
    merge_order(arr, tmp, mid, right);
    merge_step(arr, tmp, left, mid, right);
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n = 0;
    if (!(cin >> n)) {
        return 0;
    }
    if (n <= 0) {
        return 0;
    }

    vector<Item> arr(n);
    for (int i = 0; i < n; i++) {
        int v = 0;
        cin >> v;
        arr[i].value = v;
        arr[i].digit_sum = calc_digit_sum(v);
    }

    vector<Item> tmp(n);
    merge_order(arr, tmp, 0, n);

    for (int i = 0; i < n; i++) {
        if (i) {
            cout << ' ';
        }
        cout << arr[i].value;
    }
    cout << ' ';
    return 0;
}
